src/snowflake/snowpark/_internal/packaging_utils.py

```diff
@@ -0,0 +1,296 @@
+#
+# Copyright (c) 2012-2023 Snowflake Computing Inc. All rights reserved.
+#
+# The code in this file is largely a copy of https://github.com/Snowflake-Labs/snowcli/blob/main/src/snowcli/utils.py
+
+import glob
+import os
+import platform
+import subprocess
+import sys
+import zipfile
+from logging import getLogger
+from pathlib import Path
+from typing import Dict, List, Optional, Set, Tuple
+
+import pkg_resources
+from pkg_resources import Requirement
+
+_logger = getLogger(__name__)
+PIP_ENVIRONMENT_VARIABLE = "PIP_NAME"
+IMPLICIT_ZIP_FILE_NAME = "zipped_packages"
+SNOWPARK_PACKAGE_NAME = "snowflake-snowpark-python"
+
+
+def get_package_name_from_metadata(metadata_file_path: str) -> Optional[str]:
+    """Loads a METADATA file from the dist-info directory of an installed
+    Python package, finds the name of the package.
+    This is found on a line containing "Name: my_package".
+
+    Args:
+        metadata_file_path (str): The path to the METADATA file
+
+    Returns:
+        str: the name of the package.
+    """
+    import re
+
+    with open(metadata_file_path, encoding="utf-8") as metadata_file:
+        contents = metadata_file.read()
+        results = re.search("^Name: (.*)$", contents, flags=re.MULTILINE)
+        if results is None:
+            return None
+        requirement_line = results.group(1)
+        results = re.search("^Version: (.*)$", contents, flags=re.MULTILINE)
+        if results is not None:
+            version = results.group(1)
+            requirement_line += f"=={version}"
+        return requirement_line.strip().lower()
+
+
+def get_downloaded_packages(directory: str) -> Dict[Requirement, List[str]]:
+    """
+    This function records correspondence between installed python packages and their folder structure.
+    When redundant packages are detected, we can use this correspondence to delete installed python packages.
+
+    :param directory: Target folder in which pip installed the packages
+    :return: Mapping from package to a list of unique folder/file names that correspond to it.
+    """
+    import glob
+    import os
+
+    metadata_files = glob.glob(os.path.join(directory, "*dist-info", "METADATA"))
+    package_name_to_record_entries_map: Dict[Requirement, List[str]] = {}
+    for metadata_file in metadata_files:
+        parent_folder = os.path.dirname(metadata_file)
+        package = get_package_name_from_metadata(metadata_file)
+
+        if package is not None:
+            # Determine which folders or files belong to this package
+            record_file_path = os.path.join(parent_folder, "RECORD")
```

Comment:
Adding a comment here on the function of record files and record entries will be useful context

Reply:
Thanks, adding to docstring